File: code/go/robot_control.go
package main

/*
 * TEKNOFEST Go Robot Kontrol Örneği
 * Concurrent robot kontrol sistemi
 */

import "fmt"

type Direction int

const (
	Kuzey Direction = iota
	Doğu
	Güney
	Batı
)

type Robot struct {
	name      string
	x, y      int
	direction Direction
}

func NewRobot(name string) *Robot {
	return &Robot{
		name:      name,
		x:         0,
		y:         0,
		direction: Kuzey,
	}
}

func (r *Robot) MoveForward(distance int) {
	switch r.direction {
	case Kuzey:
		r.y += distance
	case Güney:
		r.y -= distance
	case Doğu:
		r.x += distance
	case Batı:
		r.x -= distance
	}
	fmt.Printf("%s %d birim ileri gitti. Konum: (%d, %d)\n", 
		r.name, distance, r.x, r.y)
}

func (r *Robot) TurnRight() {
	r.direction = (r.direction + 1) % 4
	directions := []string{"Kuzey", "Doğu", "Güney", "Batı"}
	fmt.Printf("%s sağa döndü. Yön: %s\n", r.name, directions[r.direction])
}

func (r *Robot) GetPosition() (int, int, Direction) {
	return r.x, r.y, r.direction
}

func main() {
	fmt.Println("🤖 TEKNOFEST Robot Kontrol Sistemi\n")
	
	robot := NewRobot("TEKNOFEST-Bot")
	
	robot.MoveForward(3)
	robot.TurnRight()
	robot.MoveForward(2)
	robot.TurnRight()
	robot.MoveForward(1)
	
	x, y, dir := robot.GetPosition()
	fmt.Printf("\n📍 Final Konum: (%d, %d) Yön: %d\n", x, y, dir)
}

File: code/rust/robot_control.rs
/*
 * TEKNOFEST Rust Robot Kontrol Örneği
 * Güvenli ve performanslı robot kontrol sistemi
 */

#[derive(Debug)]
enum Direction {
    Kuzey,
    Doğu,
    Güney,
    Batı,
}

struct Robot {
    name: String,
    x: i32,
    y: i32,
    direction: Direction,
}

impl Robot {
    fn new(name: String) -> Self {
        Robot {
            name,
            x: 0,
            y: 0,
            direction: Direction::Kuzey,
        }
    }
    
    fn move_forward(&mut self, distance: i32) {
        match self.direction {
            Direction::Kuzey => self.y += distance,
            Direction::Güney => self.y -= distance,
            Direction::Doğu => self.x += distance,
            Direction::Batı => self.x -= distance,
        }
        println!("{} {} birim ileri gitti. Konum: ({}, {})", 
                 self.name, distance, self.x, self.y);
    }
    
    fn turn_right(&mut self) {
        self.direction = match self.direction {
            Direction::Kuzey => Direction::Doğu,
            Direction::Doğu => Direction::Güney,
            Direction::Güney => Direction::Batı,
            Direction::Batı => Direction::Kuzey,
        };
        println!("{} sağa döndü. Yön: {:?}", self.name, self.direction);
    }
    
    fn get_position(&self) -> (i32, i32, &Direction) {
        (self.x, self.y, &self.direction)
    }
}

fn main() {
    println!("🤖 TEKNOFEST Robot Kontrol Sistemi\n");
    
    let mut robot = Robot::new("TEKNOFEST-Bot".to_string());
    
    robot.move_forward(3);
    robot.turn_right();
    robot.move_forward(2);
    robot.turn_right();
    robot.move_forward(1);
    
    println!("\n📍 Final Konum: {:?}", robot.get_position());
}

